### DISCUSSION TITLE
Uvicorn behind proxy: suggested improvements for documentation and command line options

### DISCUSSION THREAD

**Discussion by adamsol** [Kludex/uvicorn#2374] [Q&A]
I'd like to report some details that I found confusing when configuring Uvicorn with Nginx (https://www.uvicorn.org/deployment/#running-behind-nginx).

* The documentation warns to "ensure [...] that Uvicorn is run using the `--proxy-headers` setting". However, this is the default: https://github.com/encode/uvicorn/blob/d79f285184404694c77f7ca649858e7488270cf7/uvicorn/main.py#L221-L224

* So the above option does nothing, and it's not the only one. Why are there pairs of boolean options such that one of them does nothing? We don't even know what the defaults are, without searching the code.

* When binding to a UNIX socket, we need to also add `--forwarded-allow-ips='*'`. This is mentioned in the documentation, but it's quite unexpected and easy to miss. I missed it, and spent a lot of time trying to understand why my requests passed to Django are not considered secure, and looking for workarounds.
  * Shouldn't `--forwarded-allow-ips='*'` be the default when `--uds` is used? Currently passing the `*` makes me feel like I'm doing something wrong security-wise.
  * If not, then maybe at least it could be described in the documentation whenever `--uds` is described, e.g. [here](https://www.uvicorn.org/settings/#socket-binding) and in `uvicorn --help`.

**Comment by bemoody** [comment]
> Currently passing the * makes me feel like I'm doing something wrong security-wise.

I think your intuition is correct.

Using `--forwarded-allow-ips=*`, together with `proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for` as recommended, means that any client can spoof any IP address by sending an X-Forwarded-For header.

That's a problem if knowing the true IP address is significant for security.
